# 01.Importing Libraries

In [1]:
# Importing Libraries

import pandas as pd
import numpy as np
import os

# 02.Import Orders dataset from Prepared Data

In [2]:
path=r'C:\Users\LENOVO\Desktop\CF - Data Analytics\Achievement 5\Achievement 5 - PROJECT'

# Assigning path of data stored to a variable

In [7]:
df_ords=pd.read_csv(os.path.join(path, '02 Data', 'Prepared Data', 'orders_wrangled.csv'), index_col=False)

# Importing wrangled order dataset

# 03.Import Products dataset from Original Data

In [4]:
df_prods=pd.read_csv(os.path.join(path, '02 Data', 'Original Data', 'products.csv'), index_col=False)

# Importing product dataset

In [8]:
df_ords.head()

# Checking data present in wrangled order dataset

Unnamed: 0  order_id  user_id  order_number  orders_day_of_week  \
0           0   2539329        1             1                   2   
1           1   2398795        1             2                   3   
2           2    473747        1             3                   3   
3           3   2254736        1             4                   4   
4           4    431534        1             5                   4   

   ordered_hour  days_since_prior_order  
0             8                     NaN  
1             7                    15.0  
2            12                    21.0  
3             7                    29.0  
4            15                    28.0

In [6]:
df_prods.head()

# Checking data present in product dataset

product_id                                       product_name  aisle_id  \
0           1                         Chocolate Sandwich Cookies        61   
1           2                                   All-Seasons Salt       104   
2           3               Robust Golden Unsweetened Oolong Tea        94   
3           4  Smart Ones Classic Favorites Mini Rigatoni Wit...        38   
4           5                          Green Chile Anytime Sauce         5   

   department_id  prices  
0             19     5.8  
1             13     9.3  
2              7     4.5  
3              1    10.5  
4             13     4.3

# 04.Creating Test Dataframe for Mixed Type Column

In [9]:
# Create a Dataframe

df_test=pd.DataFrame()

In [10]:
# Create a mixed type column

df_test['mix']=['a', 'b', 1, True]

In [11]:
df_test.head()

mix
0     a
1     b
2     1
3  True

In [12]:
df_test['mix']=df_test['mix'].astype('str')

# Converting the datatype of 'mix' to string

In [23]:
df_test['mix'].dtype

# checking the resulting datatype

dtype('O')

# 05.Finding Missing Values

In [24]:
df_prods.isnull().sum()

# Checking for missing values in Products table

product_id        0
product_name     16
aisle_id          0
department_id     0
prices            0
dtype: int64

In [25]:
df_nan=df_prods.loc[df_prods['product_name'].isnull()==True]

# To view only the data with missing values in the Products dataframe by creating a subset

In [26]:
df_nan

# Checking the results of the subset

product_id product_name  aisle_id  department_id  prices
33             34          NaN       121             14    12.2
68             69          NaN        26              7    11.8
115           116          NaN        93              3    10.8
261           262          NaN       110             13    12.1
525           525          NaN       109             11     1.2
1511         1511          NaN        84             16    14.3
1780         1780          NaN       126             11    12.3
2240         2240          NaN        52              1    14.2
2586         2586          NaN       104             13    12.4
3159         3159          NaN       126             11    13.1
3230         3230          NaN       120             16    14.4
3736         3736          NaN        41              8    14.8
4283         4283          NaN        77              7    14.4
4790         4790          NaN        91             16    14.5
38187       38183          NaN        39             12    20.9
40444       40440          NaN       120             16    14.8

# 06.Addressing Missing Values

In [28]:
df_prods.shape

#To check how many rows are present in the Products dataframe

(49693, 5)

In [29]:
df_prods_clean=df_prods[df_prods['product_name'].isnull()==False]

# To create a subset dataframe of the Products table WITHOUT missing values

# isnull() == False, so that only rows WITHOUT missing data is added into the subset dataframe

In [30]:
df_prods_clean.shape

(49677, 5)

In [ ]:
# 49693 - 49677 = 16 rows with missing data are not created in the new subset dataframe

# 07.Finding Duplicate Values

In [31]:
df_dups=df_prods_clean[df_prods_clean.duplicated()]

In [32]:
df_dups

product_id                                       product_name  \
462           462                  Fiber 4g Gummy Dietary Supplement   
18459       18458                                         Ranger IPA   
26810       26808               Black House Coffee Roasty Stout Beer   
35309       35306  Gluten Free Organic Peanut Butter & Chocolate ...   
35495       35491                            Adore Forever Body Wash   

       aisle_id  department_id  prices  
462          70             11     4.8  
18459        27              5     9.2  
26810        27              5    13.4  
35309       121             14     6.8  
35495       127             11     9.9

# 08.Addressing Duplicate Values

In [33]:
df_prods_clean.shape

# Checking how many rows and columns are present in the subset table

(49677, 5)

In [34]:
df_prods_clean_no_dups=df_prods_clean.drop_duplicates()

# Removing duplicates and assigning the subset to a new Products dataframe

In [35]:
df_prods_clean_no_dups.shape

# Checking how many rows and columns ar epresent after removing duplicates

(49672, 5)

# 09.Exporting the Clean Dataframe

In [36]:
df_prods_clean_no_dups.to_csv(os.path.join(path, '02 Data', 'Prepared Data', 'products_checked.csv'))

# exporting cleaned Products dataframe to local folder in CSV format

# ----- EXERCISE 5.5 - DATA CONSISTENCY CHECKS ----

# Task 2

In [44]:
df_prods.describe()

# Checking all descriptive statistics about Products dataframe

product_id      aisle_id  department_id        prices
count  49693.000000  49693.000000   49693.000000  49693.000000
mean   24844.345139     67.770249      11.728433      9.994136
std    14343.717401     38.316774       5.850282    453.519686
min        1.000000      1.000000       1.000000      1.000000
25%    12423.000000     35.000000       7.000000      4.100000
50%    24845.000000     69.000000      13.000000      7.100000
75%    37265.000000    100.000000      17.000000     11.200000
max    49688.000000    134.000000      21.000000  99999.000000

## The maximum value in the Prices column is 99999. This could possibly be an error as it is a huge price for an online grocery store and needs to be investigated further.

# Task 3

In [39]:
 # Checking for mixed data type in Orders dataframe

for col in df_ords.columns.tolist():
    weird=(df_ords[[col]].applymap(type) != df_ords[[col]].iloc[0].apply(type)).any(axis=1)
    if len(df_ords[weird]) > 0:
        print(col)

## Since there was no output it implies that there is no mixed data type present in the Orders table

# Task 5

In [40]:
df_ords.isnull().sum()

# Checking for MISSING values in the Orders table

Unnamed: 0                     0
order_id                       0
user_id                        0
order_number                   0
orders_day_of_week             0
ordered_hour                   0
days_since_prior_order    206209
dtype: int64

In [41]:
df_ords.loc[df_ords['days_since_prior_order'].isnull()==True]

# To see the rows in Orders dataframe with the MISSING values

Unnamed: 0  order_id  user_id  order_number  orders_day_of_week  \
0                 0   2539329        1             1                   2   
11               11   2168274        2             1                   2   
26               26   1374495        3             1                   1   
39               39   3343014        4             1                   6   
45               45   2717275        5             1                   3   
...             ...       ...      ...           ...                 ...   
3420930     3420930    969311   206205             1                   4   
3420934     3420934   3189322   206206             1                   3   
3421002     3421002   2166133   206207             1                   6   
3421019     3421019   2227043   206208             1                   1   
3421069     3421069   3154581   206209             1                   3   

         ordered_hour  days_since_prior_order  
0                   8                     NaN  
11                 11                     NaN  
26                 14                     NaN  
39                 11                     NaN  
45                 12                     NaN  
...               ...                     ...  
3420930            12                     NaN  
3420934            18                     NaN  
3421002            19                     NaN  
3421019            15                     NaN  
3421069            11                     NaN  

[206209 rows x 7 columns]

## There are 206209 rows of missing data for the column 'days_since_prior_order'.
## This could be due to the fact these are the latest orders made on the website

# Task 6

In [45]:
df_ords.shape

# Checking the number of rows present in Orders table

(3421083, 7)

In [65]:
df_ords['days_since_prior_order'].fillna(0, inplace=True)

# Imputing 0 in place of NaN for the rows with missing values in Orders table

In [66]:
df_ords.loc[df_ords['days_since_prior_order'].isnull()==True]

# Checking to see if there are any missing values in the Orders table

Empty DataFrame
Columns: [Unnamed: 0, order_id, user_id, order_number, orders_day_of_week, ordered_hour, days_since_prior_order]
Index: []

## Missing values are replaced by 0.

# Task 7 and 8

In [51]:
df_ords_dups=df_ords[df_ords.duplicated()]

# Checking for duplicate data in the Orders table and moving the duplicate data to a new dataframe 'df_ords_dups'

In [69]:
df_ords_dups

# Printing duplicate data if present

Empty DataFrame
Columns: [Unnamed: 0, order_id, user_id, order_number, orders_day_of_week, ordered_hour, days_since_prior_order]
Index: []

## There are NO duplicate data in the Orders table

# Task 9

In [71]:
df_ords.to_csv(os.path.join(path, '02 Data', 'Prepared Data', 'orders_checked.csv'))

# exporting cleaned Orders dataframe to local folder in CSV format

# More Data Consistency Checks - was done in Ex 5.9

In [ ]:
# df_ORDS_PRODS_MERGE['prices'].describe()

# df_ORDS_PRODS_MERGE['prices'].median()

# df_ORDS_PRODS_MERGE.loc[df_ORDS_PRODS_MERGE['prices'] > 100]  
    # To check if there are prices with value above 100

# df_ORDS_PRODS_MERGE.loc[df_ORDS_PRODS_MERGE['prices'] >100, 'prices'] = np.nan  
    # Marks the price values above 100 as MISSING(NaN)
